### ISSUE TITLE
Platform Request: PowerVS (Power Virtual Server)

### ISSUE THREAD

**Issue by Prashanth684** [coreos/fedora-coreos-tracker#817]
In order to implement support for a new cloud platform in Fedora CoreOS, we need to know several things about the platform.  Please try to answer as many questions as you can.

- [ ] Why is the platform important?  Who uses it?
The Power Virtual Server platform [1] is an extension of the IBMCloud platform which offers instances of Power systems. The main requirement to support this platform is for Openshift IPI deployments [2]

- [ ] What is the official name of the platform?  Is there a short name that's commonly used in client API implementations?
PowerVS. The APIs will use `powervs`

- [ ] How can the OS retrieve instance userdata?  What happens if no userdata is provided?
uerdata is provided through config drive similar to Openstack.

- [ ] Does the platform provide a way to configure SSH keys for the instance?  How can the OS retrieve them?  What happens if none are provided?
Afterburn? will need to confirm with  the IBM PowerVS team

- [ ] How can the OS retrieve network configuration?  Is DHCP sufficient, or is there some other network-accessible metadata service?
DHCP. But let  me also confirm if any additional config is required.

- [ ] In particular, how can the OS retrieve the system hostname?
TBD

- [ ] Does the platform require the OS to have a specific console configuration?
These are power virtual machines so the console is `hvc0` as in any other ppc64le images that are generated today.

- [ ] Is there a mechanism for the OS to report to the platform that it has successfully booted?  Is the mechanism required?
TBD

- [ ] Does the platform have an agent that runs inside the instance?  Is it required?  What does it do?  What language is it implemented in, and where is the source code repository?
TBD

- [ ] How are VM images uploaded to the platform and published to other users?  Is there an API?  What disk image format is expected?
VM images are uploaded to buckets in the  IBMCloud cloud object storage. The APIs used are ibmcloud sdk and bluemix client to interface with the cloud object storage. the disk format is an ova format with a raw disk image and a custom ovf format specific to PowerVS.
 
- [ ] Are there any other platform quirks we should know about?

[1] https://www.ibm.com/cloud/power-virtual-server
[2] openshift/enhancements#736

**Comment by Prashanth684** [comment]
Some other info from #815:

 In order to support this platform we would need:
- platform ID powervs to be recognized for this platform by various components. This is mainly needed for ignition, for which this platform would make use of the Openstack way of fetching ignition config as Power Virtual servers are based on the gen1 IBM Cloud which mimics Openstack.
- a buildextend-powervs command in coreos-assembler to create powervs specific OVA images. these ova images consists of raw disk images with a particular ovf format required for Power virtual servers.
- ore upload support in coreos-assembler to upload images to ibmcloud. PowerVS would require these images to be in a storagebucket in the IBMCloud which can then be retrieved by the Openshift installer to perform IPI deploys. The ore command would apply to the IBMCloud platform rather than the PowerVS platform as the images are stored in IBMCloud.
- add the new platform powervs to the cosa schema in coreos-assembler so that the artifacts can be written into the metadata

**Comment by manojnkumar** [comment]
- [ ] In particular, how can the OS retrieve the system hostname?
With the regular methods, hostname(1) and gethostname(2)

- [ ] Is there a mechanism for the OS to report to the platform that it has successfully booted? Is the mechanism required?
The RMC agent can report that the OS has successfully booted. The mechanism is optional.

- [ ] Does the platform have an agent that runs inside the instance? Is it required? What does it do? What language is it implemented in, and where is the source code repository?
The RMC agent provides health status and heart-beat. It is required. It provides resource monitoring and control. It is part of the RSCT that is required for dynamic resize operators. I believe it is written in C (need to confirm). Source code is only available in private repos. https://www.ibm.com/docs/en/rsct/3.2?topic=administering-concepts

**Comment by bgilbert** [comment]
Thanks for filing.

>> Does the platform provide a way to configure SSH keys for the instance? How can the OS retrieve them? What happens if none are provided?
>
> Afterburn? will need to confirm with the IBM PowerVS team

Right, the question is how Afterburn should fetch them.

> custom ovf format specific to PowerVS.

What's custom about it?

>>  In particular, how can the OS retrieve the system hostname?
>
> With the regular methods, hostname(1) and gethostname(2)

The question is how the OS knows its own hostname.  For example, does the OS obtain it from the DHCP hostname field?

> The RMC agent provides health status and heart-beat. It is required. It provides resource monitoring and control. It is part of the RSCT that is required for dynamic resize operators. I believe it is written in C (need to confirm). Source code is only available in private repos.

Fedora CoreOS cannot ship a closed-source agent.  If certain agent functionality is required by the cloud and the agent cannot be open-sourced, we'd need to reimplement the mandatory functionality in Afterburn or elsewhere.  (We've done this before, since we prefer not to ship cloud agents in Fedora CoreOS.)

**Comment by Prashanth684** [comment]
> Thanks for filing.
> 
> > > Does the platform provide a way to configure SSH keys for the instance? How can the OS retrieve them? What happens if none are provided?
> > 
> > 
> > Afterburn? will need to confirm with the IBM PowerVS team
> 
> Right, the question is how Afterburn should fetch them.
> 
> > custom ovf format specific to PowerVS.
> 
> What's custom about it?
https://developer.ibm.com/powervc/2018/01/19/whats-inside-powervc-image-ova/ describes what's inside a powervs ova.
> 
> > > In particular, how can the OS retrieve the system hostname?
> > 
> > 
> > With the regular methods, hostname(1) and gethostname(2)
> 
> The question is how the OS knows its own hostname. For example, does the OS obtain it from the DHCP hostname field?
Yes. this is through DHCP.
> 
> > The RMC agent provides health status and heart-beat. It is required. It provides resource monitoring and control. It is part of the RSCT that is required for dynamic resize operators. I believe it is written in C (need to confirm). Source code is only available in private repos.
> 
> Fedora CoreOS cannot ship a closed-source agent. If certain agent functionality is required by the cloud and the agent cannot be open-sourced, we'd need to reimplement the mandatory functionality in Afterburn or elsewhere. (We've done this before, since we prefer not to ship cloud agents in Fedora CoreOS.)

**Comment by bgilbert** [comment]
Replying to https://github.com/coreos/fedora-coreos-tracker/issues/815#issuecomment-832933780:

AIUI we currently support [only IBM Cloud VPC Generation 2](https://github.com/coreos/coreos-assembler/issues/1119#issuecomment-584539165).  What are the differences between IBMCloud Gen1 x86 and PowerVS?  If they're minor, and if there's no such thing as "IBM Cloud Gen1 Power but not PowerVS", we could just call it e.g. `ibmcloud-gen1` and distinguish any remaining differences by target architecture.

**Comment by Prashanth684** [comment]
> > custom ovf format specific to PowerVS.
> 
> What's custom about it?
> 

https://developer.ibm.com/powervc/2018/01/19/whats-inside-powervc-image-ova/ describes what's inside a powervs ova.

> > > In particular, how can the OS retrieve the system hostname?
> > 
> > 
> > With the regular methods, hostname(1) and gethostname(2)
> 
> The question is how the OS knows its own hostname. For example, does the OS obtain it from the DHCP hostname field?

Yes. this is through DHCP

**Comment by Prashanth684** [comment]
> Replying to [#815 (comment)](https://github.com/coreos/fedora-coreos-tracker/issues/815#issuecomment-832933780):
> 
> AIUI we currently support [only IBM Cloud VPC Generation 2](https://github.com/coreos/coreos-assembler/issues/1119#issuecomment-584539165). What are the differences between IBMCloud Gen1 x86 and PowerVS? If they're minor, and if there's no such thing as "IBM Cloud Gen1 Power but not PowerVS", we could just call it e.g. `ibmcloud-gen1` and distinguish any remaining differences by target architecture.

true. i am not sure about the exact differences between the two and if PowerVS is any different from Gen1 itself. @manojnkumar  might know. But in any case, we would need a new platform id (at least for ignition to fetch the config) to distinguish the two was my point. is my understanding of your statement correct @bgilbert ?

**Comment by bgilbert** [comment]
@Prashanth684 Yes, and I now agree that we should use a new platform ID.

**Comment by Prashanth684** [comment]
> > Replying to [#815 (comment)](https://github.com/coreos/fedora-coreos-tracker/issues/815#issuecomment-832933780):
> > AIUI we currently support [only IBM Cloud VPC Generation 2](https://github.com/coreos/coreos-assembler/issues/1119#issuecomment-584539165). What are the differences between IBMCloud Gen1 x86 and PowerVS? If they're minor, and if there's no such thing as "IBM Cloud Gen1 Power but not PowerVS", we could just call it e.g. `ibmcloud-gen1` and distinguish any remaining differences by target architecture.
> 
> true. i am not sure about the exact differences between the two and if PowerVS is any different from Gen1 itself. @manojnkumar might know. But in any case, we would need a new platform id (at least for ignition to fetch the config) to distinguish the two was my point. is my understanding of your statement correct @bgilbert ?

After having a discussion with the IBM folks, looks like Power VS deviates a bit from IBM Cloud Gen1 and is its own offering. Looks like using `powervs` as a platform name is the way to go as it denotes `Power systems Virtual Servers` which reflects the specifics of the offering which is power systems in public cloud.

**Comment by Prashanth684** [comment]
> Okay, thanks for checking.
> 
> I think the best next step here is to develop a plan to address the mandatory closed-source agent. If we don't have a solution to that problem, we will not be able to support this platform.

yes we had a brief chat about that too and it looks like that will not be a required agent, but i will let @manojnkumar confirm.

**Comment by bpradipt** [comment]
> > Okay, thanks for checking.
> > I think the best next step here is to develop a plan to address the mandatory closed-source agent. If we don't have a solution to that problem, we will not be able to support this platform.
> 
> yes we had a brief chat about that too and it looks like that will not be a required agent, but i will let @manojnkumar confirm.

The RMC agent can be deployed as part of day-2 operation as it's being done today.  The agent is not mandatory.

**Comment by dustymabe** [comment]
We discussed this in our [meeting on 05/12/21](https://lists.fedoraproject.org/archives/list/coreos@lists.fedoraproject.org/message/O2IRLYHB7RKRMCHYCEFPVJWGUIU42CSJ/).

```
17:00:42 <dustymabe> #info powervs as a new platform seems reasonable
```

**Comment by Prashanth684** [comment]
> > > custom ovf format specific to PowerVS.
> > 
> > 
> > What's custom about it?
> 
> https://developer.ibm.com/powervc/2018/01/19/whats-inside-powervc-image-ova/ describes what's inside a powervs ova.
> 
> > > > In particular, how can the OS retrieve the system hostname?
> > > 
> > > 
> > > With the regular methods, hostname(1) and gethostname(2)
> > 
> > 
> > The question is how the OS knows its own hostname. For example, does the OS obtain it from the DHCP hostname field?
> 
> Yes. this is through DHCP

After discussing with the team - the hostname for the node will be gotten via the metadata through confi drive and set through Afterburn. the method is similar to what Openstack does today.
